### ISSUE TITLE
Remove Unix domain socket files

### ISSUE THREAD

**Issue by oberstet** [crossbario/crossbar#329]
When a node uses Unix domain sockets, and the node is killed, the socket file remains and leads to errors upon restart:

```
Mai 19 14:48:09 bvr-sql18 crossbar[24764]: 2015-05-19 14:48:09+0200 [Controller  24764] ApplicationError: ApplicationError('crossbar.error.cannot_listen', args = ("ERROR: cannot listen on transport endpoint (Couldn't listen on any:/home/adr-sqlb/crossbar/.crossbar/wamp1.sock: [Errno 98] Address already in use.)",), kwargs = {})
```

**Comment by Altern8** [comment]
While waiting for this to be fixed, I wonder if there is actually a workaround for releasing the socket file handle and allow Crossbar to restart.

**Comment by oberstet** [comment]
Yes: don't kill CB, but terminate it. It will then remove the file already today.

**Comment by hawkowl** [comment]
@oberstet what is the desired behaviour here? Crossbar removes the socket file on startup before listening on it?

**Comment by meejah** [comment]
I think the behavior should be that under the same circumstances when we determine a ``node.pid`` file is stale, any unix-socket files in the config should be considered stale also and deleted.

(There's of course and edge-case here: if a unix socket was changed/removed from the config after a hard-kill/crash but before restart. That won't cause any problems, and I think can safely be ignored).

**Comment by oberstet** [comment]
@hawkowl I think what @meejah describes would be convenient and sane behavior. alternatively: is it possible to check if a unix-socket path is currently opened (listened on) by some process?

**Comment by meejah** [comment]
I'm not sure how ``lsof`` does this, but the ``psutil`` Python module would allow you to enumerate all processes, and enumerate all open-files in those processes (if you have permissions) and thus answer that question.

However, unless we're running as root (hopefully not!), this wouldn't be very complete or reliable...shelling out to ``lsof`` would probably be somewhat more reliable, as it's likely installed suid-root (e.g. ``lsof -a -U /tmp/socket-file`` would list every process that has ``/tmp/socket-file`` open as a Unix socket. I think.) Of course, ``lsof`` isn't going to work on Windows (and may not be installed on linux-y OSes). Not sure about *BSDs.

BUT all that said: if you *do* have permission to delete a socket path, *and* you've got it in your current crossbar configuration at startup...maybe it's okay to just go ahead and (try to) delete it...?
Maybe it's less-rude to provide a ``--delete`` or similar option, and mention that in the error message? (e.g. "Can't listen on '/tmp/socket-file' either delete it or pass --delete to 'crossbar start'"). What do things like ``nginx`` do in these situations?

**Comment by oberstet** [comment]
> BUT all that said: if you do have permission to delete a socket path, and you've got it in your current crossbar configuration at startup...maybe it's okay to just go ahead and (try to) delete it...?

That makes sense.

So, which of the following?

A. Whenever CB wants to start a transport on a UDS, it first checks if the path exists, and if so, it removes the path.

or

B. When CB starts and detects that it wasn't shut down cleanly, it'll not only remove any dangling PID file, but also will try-remove any UDS socket path on any configured transport.

A. seems easier, and also works with dynamic configuration (where CB starts "empty" .. no transports at all).

So in essence, start_transport should just check-try-remove the UDS path provided in the config parameter when a UDS transport is started. Simple, and works with dyn. config as well.

**Comment by meejah** [comment]
Personally, I prefer A. Probably it should ``log.info`` that it's doing/trying that though...

I guess I was just adding caveats that I'm not 100% sure this is what other things like this do. But, it makes sense to me (and is probably what 99% of users are going to try when the warning says "try deleting the socket file" -- and so the software should try that, IMO).

**Comment by oberstet** [comment]
fixed in PR #427
